File: resumeable_upload/golang/services/file_service.go
package services

import (
	"context"
	"fmt"
	"io"
	"mime"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type FileService interface {
	Upload(ctx context.Context, fileId string, chunkIndex string, body io.ReadCloser) string
	Merge(ctx context.Context, fileId string, totalChunks int) string
	CheckFile(ctx context.Context, fileId string) ([]int, error)
	UploadAndMerge(ctx context.Context, fileId string, chunkIndex string, lastChunkIndex string, body io.ReadCloser) (response map[string]interface{})
}

type fileService struct {
}

func NewFileService() FileService {
	return &fileService{}
}

func (service *fileService) Upload(ctx context.Context, fileId string, chunkIndex string, body io.ReadCloser) string {
	temporaryPath := fmt.Sprintf("uploads/%s.part%s", fileId, chunkIndex)
	file, err := os.Create(temporaryPath)
	if err != nil {
		return err.Error()
	}
	defer file.Close()

	_, err = io.Copy(file, body)
	if err != nil {
		return err.Error()
	}

	return "success upload"
}

func (service *fileService) Merge(ctx context.Context, fileId string, totalChunks int) string {
	outputFilePatch := fmt.Sprintf("uploads/" + fileId + ".merged")
	outputFile, err := os.Create(outputFilePatch)
	if err != nil {
		return err.Error()
	}
	defer outputFile.Close()

	for i := 0; i < totalChunks; i++ {
		chunkPath := fmt.Sprintf("uploads/%s.part%d", fileId, i)
		chunkFile, err := os.Open(chunkPath)
		if err != nil {
			return err.Error()
		}
		io.Copy(outputFile, chunkFile)
		chunkFile.Close()
		os.Remove(chunkPath)
	}

	buffer := make([]byte, 512)
	outputFile.Seek(0, io.SeekStart)
	_, err = outputFile.Read(buffer)
	if err != nil {
		return err.Error()
	}
	mimeType := http.DetectContentType(buffer)
	extention := ".bin"
	extentions, err := mime.ExtensionsByType(mimeType)
	fmt.Println("extentions:", extentions)
	if err != nil {
		return err.Error()
	}
	if len(extentions) > 0 {
		extention = extentions[0]
	}
	// you can check the extention file here

	finalPath := fmt.Sprintf("uploads/%s%s", fileId, extention)
	err = os.Rename(outputFilePatch, finalPath)
	if err != nil {
		return err.Error()
	}
	return "success merge"
}

func (service *fileService) CheckFile(ctx context.Context, fileId string) ([]int, error) {
	pattern := fmt.Sprintf("uploads/%s.part*", fileId)
	matches, err := filepath.Glob(pattern)
	if err != nil {
		return []int{}, err
	}
	var indices []int
	for _, path := range matches {
		base := filepath.Base(path)
		parts := strings.Split(base, ".part")
		if len(parts) == 2 {
			index, err := strconv.Atoi(parts[1])
			if err != nil {
				return []int{}, err
			}
			indices = append(indices, index)
		}
	}
	return indices, nil
}

func (service *fileService) UploadAndMerge(ctx context.Context, fileId string, chunkIndex string, lastChunkIndex string, body io.ReadCloser) (response map[string]interface{}) {
	var err error

	// from file header multipart file header (upload by user through http) you cannot directly merge it into file, you have to save it to disk, open it and then merge it
	temporaryPath := fmt.Sprintf("uploads/%s.temp%s", fileId, chunkIndex)
	file, err := os.Create(temporaryPath)
	if err != nil {
		return map[string]interface{}{
			"response": err.Error(),
		}
	}
	defer file.Close()
	_, err = io.Copy(file, body)
	if err != nil {
		return map[string]interface{}{
			"response": err.Error(),
		}
	}

	tempFile, err := os.Open(temporaryPath)
	if err != nil {
		return map[string]interface{}{
			"response": err.Error(),
		}
	}
	defer tempFile.Close()

	var outputFile *os.File
	outputFilePatch := fmt.Sprintf("uploads/" + fileId + ".merged")
	if chunkIndex == "0" {
		// have to do this, because if you use os.open or os.create, it will just readonly
		outputFile, err = os.OpenFile(outputFilePatch, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
		if err != nil {
			return map[string]interface{}{
				"response": err.Error(),
			}
		}
	} else {
		outputFile, err = os.OpenFile(outputFilePatch, os.O_WRONLY|os.O_APPEND, 0644)
		if err != nil {
			return map[string]interface{}{
				"response": err.Error(),
			}
		}
	}
	defer outputFile.Close()

	_, err = io.Copy(outputFile, tempFile)
	if err != nil {
		fmt.Println("err copy:", err)
	}
	os.Remove(temporaryPath)
	if chunkIndex == lastChunkIndex {
		// i don't know why i cannot use outputfile above, i need to open it agait with different variable name
		outputFile2, err := os.OpenFile(outputFilePatch, os.O_RDWR|os.O_APPEND, 0644)
		if err != nil {
			return map[string]interface{}{
				"response": err.Error(),
			}
		}
		defer outputFile2.Close()

		buffer := make([]byte, 512)
		outputFile.Seek(0, io.SeekStart)
		_, err = outputFile2.Read(buffer)
		if err != nil {
			fmt.Println("err read:", err)
			return map[string]interface{}{
				"response": err.Error(),
			}
		}
		mimeType := http.DetectContentType(buffer)
		extention := ".bin"
		extentions, err := mime.ExtensionsByType(mimeType)
		fmt.Println("extentions:", extentions)
		if err != nil {
			return map[string]interface{}{
				"response": err.Error(),
			}
		}
		if len(extentions) > 0 {
			extention = extentions[1]
		}
		// you can check the extention file here

		finalPath := fmt.Sprintf("uploads/%s%s", fileId, extention)
		err = os.Rename(outputFilePatch, finalPath)
		if err != nil {
			return map[string]interface{}{
				"response": err.Error(),
			}
		}
		return map[string]interface{}{
			"response": "success",
		}
	}
	return map[string]interface{}{
		"chungIndex": chunkIndex,
	}
}

File: resumeable_upload/rust/src/services/file_service.rs
use std::{fs::{self, OpenOptions}, io::Write, path::PathBuf};

use bytes::Bytes;
use futures::StreamExt;
use regex::Regex;
use tokio::{fs::File, io::{AsyncReadExt, AsyncWriteExt}};
use tokio_stream::wrappers::ReadDirStream;

pub trait FileService {
    async fn upload(&self, file_id: &str, chunk_index: &str, chunk: Bytes) -> String;
    async fn merge(&self, file_id: &str, total_chunks: &str) -> String;
    async fn check_file(&self, file_id: &str) -> (Vec<usize>, String);
    async fn upload_and_merge(&self, file_id: String, chunk_index: String, last_chunk_index: String, chunk: Bytes) -> String;
}

pub struct FileServiceImpl {}

impl FileServiceImpl {
    pub fn new() -> FileServiceImpl {
        FileServiceImpl {  }
    }
}

impl FileService for FileServiceImpl {
    async fn upload(&self, file_id: &str, chunk_index: &str, chunk: Bytes) -> String {
        let upload_dir = PathBuf::from("uploads");
        fs::create_dir_all(&upload_dir).unwrap();
        let part_path = upload_dir.join(format!("{}.part{}", file_id, chunk_index));
        let mut file = match OpenOptions::new().create(true).write(true).open(part_path) {
            Ok(file) => file,
            Err(err) => {
                println!("error: {}", err);
                return err.to_string();
            }
        };
        match file.write_all(&chunk) {
            Ok(_) => (),
            Err(err) => {
                println!("error: {}", err);
                return err.to_string();
            }
        };
        "success".to_string()
    }

    async fn merge(&self, file_id: &str, total_chunks: &str) -> String {
        let mut files: Vec<(usize, PathBuf)> = Vec::new();
        let pattern = Regex::new(&format!(r"^{}\.part(\d+)$", regex::escape(file_id))).unwrap();
        let read_dir = match tokio::fs::read_dir("uploads").await {
            Ok(r) => r,
            Err(err) => return err.to_string(),
        };
        let mut dir_stream = ReadDirStream::new(read_dir);

        while let Some(Ok(entry)) = dir_stream.next().await {
            let file_name = entry.file_name().to_string_lossy().to_string();
            if let Some(captures) = pattern.captures(&file_name) {
                if let Some(m) = captures.get(1) {
                    if let Ok(index) = m.as_str().parse::<usize>() {
                        files.push((index, entry.path()));
                    }
                }

            }
        }
        files.sort_by_key(|(index, _)| *index);
        let mut output_file = match File::create("uploads/file_upload.merged").await {
            Ok(output_file) => output_file,
            Err(err) => {
                println!("error: {}", err);
                return err.to_string();
            }
        };
        for (_, part_path) in files {
            let mut part_file = match File::open(&part_path).await {
                Ok(part_file) => part_file,
                Err(err) => {
                    println!("error: {}", err);
                    return err.to_string();
                }
            };
            let mut buffer = Vec::new();
            match part_file.read_to_end(&mut buffer).await {
                Ok(_) => (),
                Err(err) => {
                    println!("error: {}", err);
                    return err.to_string();
                }
            };
            match output_file.write_all(&buffer).await {
                Ok(_) => (),
                Err(err) => {
                    println!("error: {}", err);
                    return err.to_string();
                }
            }
            
            match tokio::fs::remove_file(&part_path).await {
                Ok(_) => (),
                Err(err) => {
                    println!("error: {}", err);
                    return err.to_string();
                }
            }
        }

        let data = match tokio::fs::read("uploads/file_upload.merged").await {
            Ok(data) => data,
            Err(err) => {
                println!("error: {}", err);
                return err.to_string();
            }
        };
        let info = match infer::get(&data) {
            Some(info) => {
                match tokio::fs::rename("uploads/file_upload.merged", format!("{}.{}", "uploads/file_upload", info.extension())).await {
                    Ok(_) => (),
                    Err(err) => {
                        println!("error: {}", err);
                        return err.to_string();
                    }
                };
            },
            None => {
                println!("cannot get extention output file");
                return "cannot get extention output file".to_string();
            }
        };
        "success".to_string()
    }

    async fn check_file(&self, file_id: &str) -> (Vec<usize>, String) {
        let mut files: Vec<(usize, PathBuf)> = Vec::new();
        let pattern = Regex::new(&format!(r"^{}\.part(\d+)$", regex::escape(file_id))).unwrap();
        let read_dir = match tokio::fs::read_dir("uploads").await {
            Ok(r) => r,
            Err(err) => return (vec![], err.to_string()),
        };
        let mut dir_stream = ReadDirStream::new(read_dir);
        while let Some(Ok(entry)) = dir_stream.next().await {
            let file_name = entry.file_name().to_string_lossy().to_string();
            if let Some(captures) = pattern.captures(&file_name) {
                if let Some(m) = captures.get(1) {
                    if let Ok(index) = m.as_str().parse::<usize>() {
                        files.push((index, entry.path()));
                    }
                }

            }
        }
        files.sort_by_key(|(index, _)| *index);
        let mut indices: Vec<usize> = vec![];
        for (index, _s) in files {
            indices.push(index);
        };
        return (indices, "success".to_string())
    }

    async fn upload_and_merge(&self, file_id: String, chunk_index: String, last_chunk_index: String, chunk: Bytes) -> String {
        let mut output_file = match tokio::fs::OpenOptions::new().create(true).append(true).open("uploads/file_upload.merged").await {
            Ok(output_file) => output_file,
            Err(err) => {
                println!("error: {}", err);
                return err.to_string();
            }
        };
        match output_file.write_all(&chunk).await {
            Ok(_) => (),
            Err(err) => {
                println!("error: {}", err);
                return err.to_string();
            }
        };

        if chunk_index == last_chunk_index {
            let data = match tokio::fs::read("uploads/file_upload.merged").await {
                Ok(data) => data,
                Err(err) => {
                    println!("error: {}", err);
                    return err.to_string();
                }
            };
            let info = match infer::get(&data) {
                Some(info) => {
                    match tokio::fs::rename("uploads/file_upload.merged", format!("{}.{}", "uploads/file_upload", info.extension())).await {
                        Ok(_) => (),
                        Err(err) => {
                            println!("error: {}", err);
                            return err.to_string();
                        }
                    };
                },
                None => {
                    println!("cannot get extention output file");
                    return "cannot get extention output file".to_string();
                }
            };
        }
        "success".to_string()
    }
}